include/dlaf/matrix/copy.h

```diff
@@ -70,4 +75,89 @@ void copy(Matrix<const T, Source>& source, Matrix<T, Destination>& dest) {
   copy(source.distribution().localNrTiles(), LocalTileIndex(0, 0), source, LocalTileIndex(0, 0), dest);
 }
 
+/// Copy of a matrix performing data reshuffling
+///
+/// Copy @p src matrix to @p dst, re-distributing data according to the @p dst distribution.
+/// @pre src.size() == dst.size()
+/// @pre equal_process_grid(src, dst)
+/// @pre mb = min(src.blockSize().rows(), dst.blockSize().rows())
+///      src.blockSize().rows() % mb == 0
+///      dst.blockSize().rows() % mb == 0
+/// @pre nb = min(src.blockSize().cols(), dst.blockSize().cols())
+///      src.blockSize().cols() % nb == 0
+///      dst.blockSize().cols() % nb == 0
+template <class T, Device Source, Device Destination>
+void copy(Matrix<T, Source>& src,  // TODO this should be const
+          Matrix<T, Destination>& dst, comm::CommunicatorGrid grid) {
+  namespace ex = pika::execution::experimental;
+
+  DLAF_ASSERT_MODERATE(equal_size(src, dst), src.size(), dst.size());
+  DLAF_ASSERT_MODERATE(equal_process_grid(src, grid), src.commGridSize(), grid.size());
+  DLAF_ASSERT_MODERATE(equal_process_grid(dst, grid), dst.commGridSize(), grid.size());
+
+  // TODO Currently multiple tile per blocks cannot be tested, as Matrix does not support it yet.
+  DLAF_ASSERT_MODERATE(src.baseTileSize() == src.blockSize(), src.baseTileSize(), src.blockSize());
+  DLAF_ASSERT_MODERATE(dst.baseTileSize() == dst.blockSize(), dst.baseTileSize(), dst.blockSize());
+
+  // Note:
+  // From an algorithmic point of view it would be better to reason in terms of block instead of tiles,
+  // with the aim of reducing the number of communications.
+  // Current implementation reasons in terms of tiles due to a limitation for "recursively retiled"
```

Comment:
Probably it would be enough even just

```suggestion
  // Current implementation reasons in terms of tiles due to a limitation for retiled
```

Reply:
Yeah, agreed on removing the "recursively" part. It's enough to have a "simple" (i.e. only once) retiled matrix as input to this algorithm to hit the limitations.